aten/src/ATen/native/Pow.cpp

```diff
@@ -6,85 +6,64 @@
 
   auto common_dtype = at::result_type(base, exp);
-  TORCH_CHECK(at::can_cast(common_dtype, result.scalar_type()),
-           "result type ", common_dtype, " can't be cast to the desired output type ",
-           result.scalar_type());
+  build_unary_op(maybe_get_output(), base.to(common_dtype));
+}
+
+TORCH_META_FUNC2(pow, Scalar) (const Scalar base, const Tensor& exp) {
+    // This overload doesn't directly use TensorIterator. It attempts to short-circuit,
+    // but otherwise redispatches to the Tensor_Tensor overload.
+    auto dtype = at::result_type(base, exp);
+    set_output(exp.sizes(), exp.options().dtype(dtype));
+}
 
+} // namespace meta
+
+namespace native {
+
+DEFINE_DISPATCH(pow_tensor_tensor_stub);
+DEFINE_DISPATCH(pow_tensor_scalar_stub);
+
+TORCH_IMPL_FUNC(pow_Tensor_Tensor_out) (const Tensor& base, const Tensor& exp, const Tensor& out) {
+  if (exp.dim() == 0 && exp.device().is_cpu() && base.is_cuda()) {
+    at::pow_out(const_cast<Tensor&>(out), base, exp.item()); // redispatch!
+  } else {
+    pow_tensor_tensor_stub(device_type(), *this);
+  }
+}
+
+TORCH_IMPL_FUNC(pow_Tensor_Scalar_out) (const Tensor& base, const Scalar exp, const Tensor& out) {
+  auto common_dtype = at::result_type(base, exp);
   if (exp.equal(0.0)) {
-    resize_output(result, base.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, base);
+    out.fill_(1);
   } else if (exp.equal(1.0)) {
-    resize_output(result, base.sizes());
-    result.copy_(base);
-    namedinference::propagate_names(result, base);
+    out.copy_(base);
   } else {
-    auto iter = TensorIterator::unary_op(result, base.to(common_dtype));
-    pow_tensor_scalar_stub(iter.device_type(), iter, exp);
+    pow_tensor_scalar_stub(device_type(), *this, exp);
   }
-  return result;
 }
 
-Tensor& pow_out(Tensor& result, Scalar base, const Tensor& exp) {
+TORCH_IMPL_FUNC(pow_Scalar_out) (const Scalar base, const Tensor& exp, const Tensor& out) {
   if (base.isComplex() && base.toComplexDouble() == 1.0) {
-    resize_output(result, exp.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, exp);
+    out.fill_(1);
   } else if (!base.isComplex() && base.toDouble() == 1.0) {
-    resize_output(result, exp.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, exp);
+    out.fill_(1);
   } else {
-    native::pow_out(result, c10::scalar_to_tensor(base, exp.device()), exp);
+    at::pow_out(const_cast<Tensor&>(out), c10::scalar_to_tensor(base, exp.device()), exp); // redispatch!
   }
-  return result;
-}
-
-Tensor& pow_(Tensor& base, const Tensor& other) {
-  return native::pow_out(base, base, other);
-}
-
-Tensor& pow_(Tensor& base, Scalar alpha) {
-  return native::pow_out(base, base, alpha);
-}
-
-Tensor pow(const Tensor& base, const Tensor& exp) {
-  auto dtype = at::result_type(base, exp);
-  Tensor result = at::empty({0}, base.options().dtype(dtype));
```

Comment:
So this was pretty awful: the functional version of `at::native::pow` creates an empty output tensor, putting on the same device as **base**. This caused an inconsistency with how `pow` deals with mixed CPU/CUDA device inputs. 

If `base` is a CUDA tensor, and `exp` is a cpu tensor with 0 dimension thenTensorIterator is happy, because `pow` uses `build_binary_op` to construct the TensorIterator, which by default sets the `allow_cpu_scalars` flag. If you flip them though, and make `exp` the CUDA tensor, and `base` the cpu tensor with 0 dimension, then this line will put the output tensor on CPU. TensorIterator will then complain that you have an output tensor on a different device than one of the inputs.

This got picked up by the structured port, because the tensor that we pass to TensorIterator is undefined: `build_binary_op(maybe_get_output(), base, exp)`. So TensorIterator ends up putting it on the "common_device", which in this case would be CUDA.

With the structured kernel port we actually end up with an INTERNAL_ASSERT instead in [Loops.cuh](https://github.com/pytorch/pytorch/blob/6da0b94dd84d16f251b181ac9da0e8715f1fff57/aten/src/ATen/native/cuda/Loops.cuh#L94), because the `pow(Tensor, Tensor)` kernel isn't written to directly deal with scalars. Updating it to use `gpu_kernel_with_scalars` instead of `gpu_kernel` fixes the issue though.

Reply:
Possibly related? #50121